lib/rubocop/cop/rspec/leading_subject.rb

```diff
@@ -3,64 +3,89 @@
 module RuboCop
   module Cop
     module RSpec
-      # Checks for `subject` definitions that come after `let` definitions.
+      # Enforce `subject` being the first definition in test.
       #
       # @example
       #   # bad
-      #   RSpec.describe User do
       #     let(:params) { blah }
       #     subject { described_class.new(params) }
       #
-      #     it 'is valid' do
-      #       expect(subject.valid?).to be(true)
-      #     end
-      #   end
+      #     before { do_something }
+      #     subject { described_class.new(params) }
       #
-      #   # good
-      #   RSpec.describe User do
+      #     it { expect_something }
       #     subject { described_class.new(params) }
+      #     it { expect_something_else }
       #
+      #
+      #   # good
+      #     subject { described_class.new(params) }
       #     let(:params) { blah }
       #
-      #     it 'is valid' do
-      #       expect(subject.valid?).to be(true)
-      #     end
-      #   end
+      #   # good
+      #     subject { described_class.new(params) }
+      #     before { do_something }
+      #
+      #   # good
+      #     subject { described_class.new(params) }
+      #     it { expect_something }
+      #     it { expect_something_else }
+      #
       class LeadingSubject < Cop
         include RangeHelp
 
-        MSG = 'Declare `subject` above any other `let` declarations.'.freeze
+        MSG = 'Declare `subject` above any other `%<offending>s` ' \
+          'declarations.'.freeze
 
         def_node_matcher :subject?, Subject::ALL.block_pattern
+        def_node_matcher :let?, Helpers::ALL.block_pattern
+        def_node_matcher :hook?, Hooks::ALL.block_pattern
+        def_node_matcher :include?, <<-PATTERN
```

Comment:
Hm. I don't know if this is just me or inconsistent with the principles here, but I frequently put `include_context` before my subject, something roughly like:

```ruby
RSpec.describe FooController do
  include_context 'controller setup'

  subject(:controller) { ... }

  # ...
end
```

I think this feels natural to me because it's analogous to the way I typically write classes

```ruby
class Foo
  include WhateverImInheriting
  
  def initialize
   # ...
  end

  # ...

end
```

where I always want to have my constructor at the "top" but inheritance comes before that.

I'm not sure if it's exactly the same, but it feels different than the other forms of setup like  lets and hooks. WDYT?

cc @bquorning

Reply:
I'm fine with allowing include_context to be on the top. What about include_examples?

Reply:
On the one hand, I think `include_examples` generally belongs below the subject. On the other hand `include_examples` is just an alias for `include_context` so it feels sort of weird to have one but not the other.

Reply:
The semantics are different; that's why we enforce shared_context or shared_examples depending whether examples are defined in the group ;)

Reply:
The expressed intent is different, yeah, but not the behavior. While I do like that we try to enforce the naming, we can't enforce the choice of `include_context` v.s. `include_examples` within a test because we can't tell if it is referring to examples or setup context or some fuzzy mix of both that maybe deserves a different term entirely. Since the proper choice of `include_*` is not enforced, it feels weird to base an additional rule off of that. I expect that this would cause some amount of frustration to people who don't use one or the other consistently, but I'm not positive.

My gut feeling is leaving it to helpers/hooks/examples is possibly a better starting place. Feel free to disagree though, I'm not firm on this.